File: src/main/java/alfheim/client/model/entity/ModelButterfly.java
package alfheim.client.model.entity;

import net.minecraft.client.model.ModelBase;
import net.minecraft.client.model.ModelRenderer;
import net.minecraft.entity.Entity;
import net.minecraft.util.MathHelper;

/**
 * ModelButterfly - AlexSocol
 * Created using Tabula 4.1.1
 */
public class ModelButterfly extends ModelBase {
    
    public ModelRenderer LeftWing;
    public ModelRenderer RightWing;
    public ModelRenderer Antenas;
    public ModelRenderer Body;

    public ModelButterfly() {
        this.textureWidth = 64;
        this.textureHeight = 32;
        this.Body = new ModelRenderer(this, 0, 0);
        this.Body.setRotationPoint(-2.0F, 16.0F, -2.0F);
        this.Body.addBox(0.0F, 0.0F, 0.0F, 4, 3, 8, 0.0F);
        this.RightWing = new ModelRenderer(this, 32, -2);
        this.RightWing.setRotationPoint(-2.0F, 17.0F, 0.0F);
        this.RightWing.addBox(0.0F, -6.0F, -1.0F, 0, 18, 16, 0.0F);
        this.setRotateAngle(RightWing, 1.5707963267948966F, 0.0F, -1.5707963267948966F);
        this.Antenas = new ModelRenderer(this, 26, 0);
        this.Antenas.setRotationPoint(-2.0F, 16.0F, -2.0F);
        this.Antenas.addBox(-2.0F, -1.0F, 0.0F, 8, 8, 0, 0.0F);
        this.setRotateAngle(Antenas, -2.356194490192345F, 0.0F, 0.0F);
        this.LeftWing = new ModelRenderer(this, 0, -2);
        this.LeftWing.setRotationPoint(2.0F, 17.0F, 0.0F);
        this.LeftWing.addBox(0.0F, -6.0F, -1.0F, 0, 18, 16, 0.0F);
        this.setRotateAngle(LeftWing, 1.5707963267948966F, 0.0F, 1.5707963267948966F);
    }

    @Override
    public void render(Entity entity, float f, float f1, float f2, float f3, float f4, float f5) {
        super.render(entity, f, f1, f2, f3, f4, f5);
        setRotationAngles(f, f1, f2, f3, f4, f5, entity);
        this.Body.render(f5);
        this.RightWing.render(f5);
        this.Antenas.render(f5);
        this.LeftWing.render(f5);
    }

    /**
     * This is a helper function from Tabula to set the rotation of model parts
     */
    public void setRotateAngle(ModelRenderer modelRenderer, float x, float y, float z) {
        modelRenderer.rotateAngleX = x;
        modelRenderer.rotateAngleY = y;
        modelRenderer.rotateAngleZ = z;
    }
    
    @Override
    public void setRotationAngles(float f, float f1, float f2, float f3, float f4, float f5, Entity entity) {
        super.setRotationAngles(f, f1, f2, f3, f4, f5, entity);
        
        RightWing.rotateAngleZ = -(MathHelper.cos(f2 * 1.7F) * (float)Math.PI * 0.5F);
        LeftWing.rotateAngleZ = MathHelper.cos(f2 * 1.7F) * (float)Math.PI * 0.5F;
    }
}

File: src/main/java/alfheim/client/model/entity/ModelButterfly.kt
package alfheim.client.model.entity

import alexsocol.asjlib.F
import net.minecraft.client.model.*
import net.minecraft.entity.Entity
import net.minecraft.util.MathHelper

/**
 * ModelButterfly - AlexSocol
 * Created using Tabula 4.1.1
 */
class ModelButterfly(val pass: Int): ModelBase() {
	
	var shape1: ModelRenderer
	var shape2: ModelRenderer
	var shape3: ModelRenderer
	var shape4: ModelRenderer
	
	init {
		textureWidth = 28
		textureHeight = 16
		shape1 = ModelRenderer(this, 0, 0)
		shape1.setRotationPoint(0.0f, 0.0f, 0.0f)
		shape1.addBox(-1.0f, -3.0f, -1.0f, 2, 10, 2, 0.0f)
		setRotateAngle(shape1, 1.5707963267948966f, 0.0f, 0.0f)
		shape2 = ModelRenderer(this, -8, 0)
		shape2.setRotationPoint(0.0f, 0.0f, 0.0f)
		shape2.addBox(1.0f, 0.0f, -4.0f, 10, 0, 16, 0.0f)
		shape3 = ModelRenderer(this, -8, 0)
		shape3.mirror = true
		shape3.setRotationPoint(0.0f, 0.0f, 0.0f)
		shape3.addBox(-11.0f, 0.0f, -4.0f, 10, 0, 16, 0.0f)
		shape4 = ModelRenderer(this, -4, 12)
		shape4.setRotationPoint(0.0f, 0.0f, 0.0f)
		shape4.addBox(-2.0f, 0.0f, -7.0f, 4, 0, 4, 0.0f)
		setRotateAngle(shape4, -0.32288591161895097f, 0.0f, 0.0f)
	}
	
	override fun render(entity: Entity, f: Float, f1: Float, f2: Float, f3: Float, f4: Float, f5: Float) {
		setRotationAngles(f, f1, f2, f3, f4, f5, entity)
		
		if (pass == 0) {
			shape1.render(f5)
			shape4.render(f5)
		} else {
			shape2.render(f5)
			shape3.render(f5)
		}
	}
	
	/**
	 * This is a helper function from Tabula to set the rotation of model parts
	 */
	fun setRotateAngle(modelRenderer: ModelRenderer, x: Float, y: Float, z: Float) {
		modelRenderer.rotateAngleX = x
		modelRenderer.rotateAngleY = y
		modelRenderer.rotateAngleZ = z
	}
	
	override fun setRotationAngles(f: Float, f1: Float, f2: Float, f3: Float, f4: Float, f5: Float, entity: Entity) {
		super.setRotationAngles(f, f1, f2, f3, f4, f5, entity)
		shape2.rotateAngleZ = MathHelper.cos(f2) * Math.PI.F * -0.25f
		shape3.rotateAngleZ = MathHelper.cos(f2) * Math.PI.F * 0.25f
	}
}